server/src/main/java/io/crate/planner/consumer/RelationNameCollector.java

```diff
@@ -60,4 +61,12 @@ public Void visitMatchPredicate(MatchPredicate matchPredicate, Set<RelationName>
         }
         return null;
     }
+
+    @Override
+    public Void visitSelectSymbol(SelectSymbol selectSymbol, Set<RelationName> context) {
+        if (selectSymbol.isCorrelated()) {
+            context.clear();
```

Comment:
If I understand correctly, context is shared throughout the visitors call chain.

Do we want to clean up the whole context if we meet at least one correlated subquery?

Asking since it seems like order of filters matters for this visitor and sometimes we clear the context and sometimes not. I replaced this by 
```
@Override
public Void visitSelectSymbol(SelectSymbol selectSymbol, Set<RelationName> context) {
    if (context.size() > 0) {
        throw new IllegalStateException("EXISTS Function going to clear non-empty context!");
    }
    return null;
}
```
and run a new `PushDownTest` case.

It passes with current version ==> context was never actually cleared, `clear()` is always called on empty set. because EXISTS gets called first. 

It fails with `IllegalStateException`^ if I swap` a.country = 'DE'` and `EXISTS`, since `EXISTS` gets called second and actually clears non-empty context.

However, both swapped and current versions passes if I leave `visitSelectSymbol` as is. What's not clear to me is that test passes regardless of the country=DE/EXISTS order even though the order seemingly affects `RelationNameCollector`s behavior. 

If clearing/not clearing the context is not important we can maybe just replace `visitSelectSymbol`  to `return null`? At least it passes locally with such version with both versions of the ordering.

Reply:
> If clearing/not clearing the context is not important we can maybe just replace visitSelectSymbol to return null? At least it passes locally with such version with both versions of the ordering.

The default implementation is: return null.
But this causes problems when a query containing a `Reference` and a `SelectSymbol` is used as only the relation name of the `Reference` will be collected, causing the `QuerySplitter` to treat the whole query as it would belong to the reference's relation, and thus the rules will push the whole query down.
E.g. this query will fail as it will falsely push down the query: https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/integrationtests/CorrelatedSubqueryITest.java#L222

Another failing example:
```
            SELECT a.mountain
            FROM sys.summits a
            WHERE
                a.height =
                (
                    SELECT height
                    FROM sys.summits b
                    WHERE
                        b.height = a.height
                )
                AND
                a.country = 'DE'
```

In this case it would falsely push down the whole query to the `a` relation.

But I agree that clearing the relation name set isn't optimal, I'll try to come up with a better solution, e.g. give the correlated SelectSymbol a virtual relation name.

Reply:
I agree with trying with a more elegant approach, but if turns out to be significantly more complicated, imho, it's also fine with clearing the context and have a comment explaining why that's needed.

Reply:
thanks for the explanation!